### PR TITLE
Add support for memory-mapped indices generation

### PR DESCRIPTION
**Pull request by victor-paltz**
- It is now possible to create an IVF,Flat index when it is optimal
- Added should_be_memory_mappable boolean to create only indices that can be memory-mapped
- Fix score_index output
- Speed up hyperparameter tuning
- Add more tests
- Add more doc

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,6 +65,27 @@ build_index(embeddings="embeddings", index_path="my_index_folder/knn.index",
...
+
+To load an index in memory mapping mode, use the following code:
+```python
+import faiss
+faiss.read_index("my_index_folder/knn.index", faiss.IO_FLAG_MMAP | faiss.IO_FLAG_READ_ONLY)
+```
+
+You can have a look to the [examples](examples/memory_mapped_autofaiss.py) to see how to use it.
+
+Technical note: You can create a direct map on IVF indices with .make_direct_map() in order to speed up
```
Comment (rom1504) [inline README.md:84]: probably mention here the option to do that that martin added in autofaiss

File: autofaiss/external/optimize.py [hunk 2]
```diff
@@ -154,25 +156,54 @@ def get_optimal_index_keys_v2(
...
-        return [f"HNSW{min(m_hnsw, 32)}"]
+    if not should_be_memory_mappable:
+        # If we can build an HNSW with the given memory constraints, it's the best
+        m_hnsw = int(floor((max_size_in_bytes / (4 * nb_vectors) - dim_vector) / 2))
+        if m_hnsw >= 8:
+            return [f"HNSW{min(m_hnsw, 32)}"]
+
+    # Try to build a not quantized IVF index
+    index_keys = get_optimal_ivf(nb_vectors)
+    index_metadata = IndexMetadata(index_keys[0], nb_vectors, dim_vector, make_direct_map=False)
```
Comment (rom1504) [inline autofaiss/external/optimize.py:182]: why make direct map hard coded to false ?

File: autofaiss/external/quantize.py [hunk 3]
```diff
@@ -49,6 +49,7 @@ def build_index(
     metric_type: str = "ip",
     nb_cores: Optional[int] = None,
     make_direct_map: bool = False,
+    should_be_memory_mappable: bool = False,
```
Comment (rom1504) [inline autofaiss/external/quantize.py:52]: can you add this to the doc ?

File: autofaiss/external/optimize.py (hunk 2 above)
Comment (victor-paltz) [reply autofaiss/external/optimize.py:182]: Just because this function doesn't take make_direct_map as parameter
It is free to add it so let's do it

**Comment by rom1504** [comment]
left a few comments on details, overall lgtm!

File: autofaiss/external/quantize.py (hunk 3 above)
Comment (victor-paltz) [reply autofaiss/external/quantize.py:52]: Ah yes, I just modified the readme

File: README.md (hunk 1 above)
Comment (victor-paltz) [reply README.md:84]: Yep!

File: autofaiss/external/quantize.py (hunk 3 above)
Comment (rom1504) [reply autofaiss/external/quantize.py:52]: can you add the parameter in the list of parameter in the readme ?

Comment (victor-paltz) [reply autofaiss/external/quantize.py:52]: Done! The doc might be a bit too much duplicated^^

**Pull request merged**